File: sample/login-panel/src/Plugins/UserService/APIs/RegisterUserMessage.ts
/**
 * RegisterUserMessage
 * desc: 用户注册，生成唯一的用户ID，用于用户注册
 * @param idCard: String (用户身份证，用于验证唯一性)
 * @param phoneNumber: String (用户手机号，用于注册及验证)
 * @param name: String (用户姓名)
 * @param password: String (用户密码，用于后续登录验证)
 * @return userID: String (注册生成的唯一用户ID)
 */
import { TongWenMessage } from 'Plugins/TongWenAPI/TongWenMessage'



export class RegisterUserMessage extends TongWenMessage {
    constructor(
        public  idCard: string,
        public  phoneNumber: string,
        public  name: string,
        public  password: string
    ) {
        super()
    }
    getAddress(): string {
        return "127.0.0.1:10012"
    }
}

File: sample/UserService/src/main/scala/APIs/UserService/RegisterUserMessage.scala
package APIs.UserService

import Common.API.API
import Global.ServiceCenter.UserServiceCode

import io.circe.{Decoder, Encoder, Json}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import io.circe.syntax.*
import io.circe.parser.*
import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

import com.fasterxml.jackson.core.`type`.TypeReference
import Common.Serialize.JacksonSerializeUtils

import scala.util.Try

import org.joda.time.DateTime
import java.util.UUID


/**
 * RegisterUserMessage
 * desc: 用户注册，生成唯一的用户ID，用于用户注册
 * @param idCard: String (用户身份证，用于验证唯一性)
 * @param phoneNumber: String (用户手机号，用于注册及验证)
 * @param name: String (用户姓名)
 * @param password: String (用户密码，用于后续登录验证)
 * @return userID: String (注册生成的唯一用户ID)
 */

case class RegisterUserMessage(
  idCard: String,
  phoneNumber: String,
  name: String,
  password: String
) extends API[String](UserServiceCode)



case object RegisterUserMessage{
    
  import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

  // Circe 默认的 Encoder 和 Decoder
  private val circeEncoder: Encoder[RegisterUserMessage] = deriveEncoder
  private val circeDecoder: Decoder[RegisterUserMessage] = deriveDecoder

  // Jackson 对应的 Encoder 和 Decoder
  private val jacksonEncoder: Encoder[RegisterUserMessage] = Encoder.instance { currentObj =>
    Json.fromString(JacksonSerializeUtils.serialize(currentObj))
  }

  private val jacksonDecoder: Decoder[RegisterUserMessage] = Decoder.instance { cursor =>
    try { Right(JacksonSerializeUtils.deserialize(cursor.value.noSpaces, new TypeReference[RegisterUserMessage]() {})) } 
    catch { case e: Throwable => Left(io.circe.DecodingFailure(e.getMessage, cursor.history)) }
  }
  
  // Circe + Jackson 兜底的 Encoder
  given registerUserMessageEncoder: Encoder[RegisterUserMessage] = Encoder.instance { config =>
    Try(circeEncoder(config)).getOrElse(jacksonEncoder(config))
  }

  // Circe + Jackson 兜底的 Decoder
  given registerUserMessageDecoder: Decoder[RegisterUserMessage] = Decoder.instance { cursor =>
    circeDecoder.tryDecode(cursor).orElse(jacksonDecoder.tryDecode(cursor))
  }


}

